src/ring_theory/graded_ring/basic.lean

```diff
@@ -0,0 +1,413 @@
+-/
+import algebra.direct_sum.algebra
+import data.mv_polynomial
+import ring_theory.polynomial.homogeneous
+
+/-! # Typeclass for graded ring
+For definition of a ring `R` being graded by `A : ι → add_subgroup R`, see doc string of 
+`graded_ring`.
+
+- `graded_ring.decompose : R → ⨁ i, A i` and `graded_ring.recompose : ⨁ i, A i → R` are the ring
+isomorphism between `R` and `⨁ i, A i` if `R` is graded by `A`.
+- `graded_ring.complete_lattice.independent` states that the `A` is independent in the sense that
+for any `i : ι`, `A i` and `⨆ (j ≠ i), A j` intersect trivially. The most noticable consequence of
+this is that `A i` and `A j` intersects trivally for distinct `i` and `j`.
+- `graded_ring.proj R A i r` is the degree `i : ι` component of `r : R`.
+- `graded_ring.support R A r` is the `finset ι` containing the `i : ι` such that the degree `i`
+component of `r` is not zero.
+- `is_homogeneous R A x` states that `x ∈ A i` for some `i : ι`.
+- `homogeneous_element R A` is the subtype of `R` of all `x` such that `is_homogeneous R A x`.
+- `mv_polynomial_is_graded` provides an instance saying that `mv_polynomial R σ` is ring graded by
+its homogeneous components.
+-/
+
+open_locale direct_sum big_operators
+
+section graded_ring
+variables (R : Type*) [ring R] {ι : Type*} (A : ι → add_subgroup R)
+  [decidable_eq ι] [add_comm_monoid ι]
+
+/-- A graded ring is a `ring R` such that `R` can be decomposed into a collection of
+  `add_subgroups R` indexed by `ι` such that the connonical map `R → ⨁ i, A i` is a bijective map
+  respecting multiplication, i.e. product of an element of degree `i` and an element of degree `j`
+  is an element of degree `i + j`.
+-/
+class graded_ring :=
+( decompose : R → ⨁ i, A i)
+( left_inv : function.left_inverse decompose (direct_sum.add_subgroup_coe A) )
+( right_inv : function.right_inverse decompose (direct_sum.add_subgroup_coe A) )
+( one_degree_zero : (1 : R) ∈ A 0 )
+( mul_respect_grading : ∀ {i j : ι} {a b : R}, a ∈ A i → b ∈ A j → a * b ∈ A (i + j))
+
+variable [graded_ring R A]
+
+lemma graded_ring.is_internal : direct_sum.add_subgroup_is_internal A :=
+⟨graded_ring.left_inv.injective, graded_ring.right_inv.surjective⟩
+
+instance gsemiring.of_ring_is_internally_graded :
+  direct_sum.gsemiring (λ i, A i) :=
+direct_sum.gsemiring.of_add_subgroups A
+  (graded_ring.one_degree_zero) (λ i j ai aj, graded_ring.mul_respect_grading ai.2 aj.2)
+
+/--If `R` is graded by `ι` with degree `i` component `A i`, then `(⨁ i, A i ≃+* R)`-/
+def graded_ring.recompose : (⨁ i, A i) ≃+* R :=
+let f : (⨁ i, A i) →+* R :=
+  direct_sum.to_semiring (λ i, (A i).subtype) rfl (λ _ _ _ _, rfl) in
+{ to_fun := f,
+  inv_fun := graded_ring.decompose,
+  left_inv := graded_ring.left_inv,
+  right_inv := graded_ring.right_inv,
+  map_mul' := ring_hom.map_mul _,
+  map_add' := ring_hom.map_add _, }
+
+lemma graded_ring.trivial_inter (i : ι) :
+   disjoint (A i) (Sup {a | ∃ j ≠ i, A j = a}) :=
```

Comment:
This should fall out in a line or two from  #10108. If it doesn't, there's a lemma missing from the API of `complete_lattice.independent`.